src/error.rs

```diff
@@ -0,0 +1,154 @@
+
+    #[error("Division by zero")]
+    DivisionByZero(NodePtr),
+
+    #[error("Divmod by zero")]
+    DivmodByZero(NodePtr),
+
+    #[error("Mod by zero")]
+    ModByZero(NodePtr),
+
+    #[error("ModPow with 0 Modulus")]
+    ModPowZeroModulus(NodePtr),
+
+    #[error("Value Stack Limit Reached")]
+    ValueStackLimitReached(NodePtr),
+
+    #[error("Environment Stack Limit Reached")]
+    EnvironmentStackLimitReached(NodePtr),
+
+    #[error("Shift too large")]
+    ShiftTooLarge(NodePtr),
+
+    #[error("Reserved operator")]
+    Reserved(NodePtr),
+
+    #[error("Invalid Operator")]
+    Invalid(NodePtr),
+
+    #[error("Unimplemented Operator")]
+    Unimplemented(NodePtr),
+
+    #[error("Operator Error: InvalidArg: {1}")]
+    InvalidArg(NodePtr, String),
+
+    #[error("CoinID Error: Invalid Parent Coin ID, not 32 bytes")]
+    CoinIDParentCoinIdNot32Bytes(NodePtr),
+
+    #[error("CoinID Error: Invalid Puzzle Hash, not 32 bytes")]
+    CoinIDPuzzleHashNot32Bytes(NodePtr),
+
+    #[error("CoinID Error: Invalid Amount: Amount is Negative")]
+    CoinIDAmountNegative(NodePtr),
+
+    #[error("CoinID Error: Invalid Amount: Amount has leading zeroes")]
+    CoinIDAmountLeadingZeroes(NodePtr),
+
+    #[error("CoinID Error: Invalid Amount: Amount exceeds max coin amount")]
+    CoinIDAmountExceedsMaxCoinAmount(NodePtr),
+
+    #[error("Allocator Error: {0}")]
+    Allocator(#[from] AllocatorErr),
+}
+impl From<std::io::Error> for EvalErr {
+    fn from(_: std::io::Error) -> Self {
+        EvalErr::SerializationError
+    }
+}
+
+impl EvalErr {
+    fn node(&self) -> Option<NodePtr> {
+        match self {
+            EvalErr::Raise(node) => Some(*node),
+            EvalErr::InvalidNilTerminator(node) => Some(*node),
+            EvalErr::DivisionByZero(node) => Some(*node),
+            EvalErr::DivmodByZero(node) => Some(*node),
+            EvalErr::ModByZero(node) => Some(*node),
+            EvalErr::ModPowZeroModulus(node) => Some(*node),
+            EvalErr::ShiftTooLarge(node) => Some(*node),
+            EvalErr::ValueStackLimitReached(node) => Some(*node),
+            EvalErr::EnvironmentStackLimitReached(node) => Some(*node),
+            EvalErr::InternalError(node, _) => Some(*node),
+            EvalErr::Reserved(node) => Some(*node),
+            EvalErr::Invalid(node) => Some(*node),
+            EvalErr::Unimplemented(node) => Some(*node),
+            EvalErr::CoinIDPuzzleHashNot32Bytes(node) => Some(*node),
+            EvalErr::CoinIDAmountNegative(node) => Some(*node),
+            EvalErr::CoinIDAmountLeadingZeroes(node) => Some(*node),
+            EvalErr::CoinIDAmountExceedsMaxCoinAmount(node) => Some(*node),
+            EvalErr::CoinIDParentCoinIdNot32Bytes(node) => Some(*node),
+            EvalErr::Allocator(alloc) => AllocatorErr::node(alloc),
+            EvalErr::InvalidArg(node, _) => Some(*node),
+            _ => None,
+        }
+    }
+    pub fn combined_str(&self) -> String {
```

Comment:
I'm still sceptical of the usability of this function. It seems like you only use it in the wasm binding now. I think just using the `Debug` formatter or the existing `to_string()` would be better.

Reply:
sounds good ill just switch back to to_string.